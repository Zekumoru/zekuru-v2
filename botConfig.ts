export default {
  messageCharactersLimit: 2000,
};
