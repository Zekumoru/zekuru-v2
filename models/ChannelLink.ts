import { Schema, Types, model } from 'mongoose';
import { ITranslateChannel } from './TranslateChannel';

interface IChannelLinkSchema {
  id: string;
  guildId: string;
  links: ITranslateChannel[];
  createdAt: Date;
}

export interface IChannelLink extends IChannelLinkSchema {
  _id: Types.ObjectId;
}

const ChannelLinkSchema = new Schema<IChannelLinkSchema>({
  id: {
    type: String,
    required: true,
  },
  guildId: {
    type: String,
    required: true,
  },
  links: [
    {
      type: Types.ObjectId,
      ref: 'TranslateChannel',
    },
  ],
  createdAt: {
    type: Date,
    default: Date.now,
  },
});

export default model('ChannelLink', ChannelLinkSchema);
