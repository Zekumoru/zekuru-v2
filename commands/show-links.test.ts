import botConfig from '../botConfig';
import {
  clearSampleChannelLinks,
  sampleChannelLinks,
  sampleTranslateChannels,
} from '../cache/__mocks__/sample-data';
import createMockInteraction from '../test/createMockChatInputCommandInteraction';
import showLinks from './show-links';
import ChannelLink from '../db/models/ChannelLink';

jest.mock('../cache');
jest.mock('../db/models/ChannelLink.ts');
jest.mock(
  '../botConfig.ts',
  () =>
    ({
      messageCharactersLimit: 250,
    } as typeof botConfig)
);

describe('/show-links command', () => {
  const mockChannelLinks = jest.mocked(ChannelLink);

  afterEach(() => {
    clearSampleChannelLinks();
    jest.clearAllMocks();
  });

  it(`should notify user that this command is only for servers`, async () => {
    const interaction = createMockInteraction({ guildId: undefined });

    await showLinks.execute(interaction);

    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
  });

  it(`should notify user if the translate channel they provided doesn't have any links yet`, async () => {
    const interaction = createMockInteraction();
    interaction.options.getChannel.mockReturnValue({ id: '100' } as any);

    await showLinks.execute(interaction);

    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
  });

  it(`should show links of a translate channel`, async () => {
    const interaction = createMockInteraction();
    sampleChannelLinks[0].links = [
      sampleTranslateChannels[1] as any,
      sampleTranslateChannels[2] as any,
    ];
    interaction.options.getChannel.mockReturnValue({ id: '100' } as any);

    await showLinks.execute(interaction);

    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
  });

  it(`should show all links and use embeds if above Discord message's characters limit`, async () => {
    const interaction = createMockInteraction();
    sampleChannelLinks[0].links = [sampleTranslateChannels[1] as any];
    sampleChannelLinks[1].links = [sampleTranslateChannels[0] as any];
    sampleChannelLinks[2].links = [sampleTranslateChannels[3] as any];
    sampleChannelLinks[3].links = [sampleTranslateChannels[2] as any];
    const guildId = interaction.guildId;
    mockChannelLinks.find.mockReturnThis();
    mockChannelLinks.populate.mockResolvedValue(sampleChannelLinks as any);
    interaction.options.getChannel.mockReturnValue(null);

    await showLinks.execute(interaction);

    expect(mockChannelLinks.find).toHaveBeenCalledWith({ guildId });
    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
    expect(interaction.reply.mock.calls[0][0]).toHaveProperty('embeds');
  });
});
