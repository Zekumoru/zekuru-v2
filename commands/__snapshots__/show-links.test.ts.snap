// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/show-links command should notify user if the translate channel they provided doesn't have any links yet 1`] = `
{
  "content": "<#100> **(en)** is linked to 0 channels: ",
}
`;

exports[`/show-links command should notify user that this command is only for servers 1`] = `
{
  "content": "This command is only available on servers.",
}
`;

exports[`/show-links command should show all links and use embeds if above Discord message's characters limit 1`] = `
{
  "embeds": [
    {
      "color": 39423,
      "description": "**Showing all translate channels links**
There are a total of 4 channels that have links.

<#100> **(en)** is linked to 1 channels: 
- <#200> **(ja)**

<#200> **(ja)** is linked to 1 channels: 
- <#100> **(en)**

<#300> **(zh)** is linked to 1 channels: 
- <#400> **(ko)**

<#400> **(ko)** is linked to 1 channels: 
- <#300> **(zh)**",
    },
  ],
}
`;

exports[`/show-links command should show links of a translate channel 1`] = `
{
  "content": "<#100> **(en)** is linked to 2 channels: 
- <#200> **(ja)**
- <#300> **(zh)**",
}
`;
